# DATA CLEANING

## Importing Necessary libraries

In [1]:
import pandas as pd
import numpy as np
from datetime import datetime
import seaborn as sns
import matplotlib.pyplot as plt

## Loading dataset and displaying max columns

In [2]:
#Loading the datset
data = pd.read_csv('fifa21 raw data v2.csv', low_memory=False)

#Display all columns 
pd.set_option('display.max_columns', None)

data.head()

ID               Name                      LongName  \
0  158023           L. Messi                  Lionel Messi   
1   20801  Cristiano Ronaldo  C. Ronaldo dos Santos Aveiro   
2  200389           J. Oblak                     Jan Oblak   
3  192985       K. De Bruyne               Kevin De Bruyne   
4  190871          Neymar Jr    Neymar da Silva Santos Jr.   

                                           photoUrl  \
0  https://cdn.sofifa.com/players/158/023/21_60.png   
1  https://cdn.sofifa.com/players/020/801/21_60.png   
2  https://cdn.sofifa.com/players/200/389/21_60.png   
3  https://cdn.sofifa.com/players/192/985/21_60.png   
4  https://cdn.sofifa.com/players/190/871/21_60.png   

                                           playerUrl Nationality  Age  ↓OVA  \
0  http://sofifa.com/player/158023/lionel-messi/2...   Argentina   33    93   
1  http://sofifa.com/player/20801/c-ronaldo-dos-s...    Portugal   35    92   
2  http://sofifa.com/player/200389/jan-oblak/210006/    Slovenia   27    91   
3  http://sofifa.com/player/192985/kevin-de-bruyn...     Belgium   29    91   
4  http://sofifa.com/player/190871/neymar-da-silv...      Brazil   28    91   

   POT                         Club     Contract   Positions Height Weight  \
0   93         \n\n\n\nFC Barcelona  2004 ~ 2021  RW, ST, CF  170cm   72kg   
1   92             \n\n\n\nJuventus  2018 ~ 2022      ST, LW  187cm   83kg   
2   93      \n\n\n\nAtlético Madrid  2014 ~ 2023          GK  188cm   87kg   
3   91      \n\n\n\nManchester City  2015 ~ 2023     CAM, CM  181cm   70kg   
4   91  \n\n\n\nParis Saint-Germain  2017 ~ 2022     LW, CAM  175cm   68kg   

  Preferred Foot  BOV Best Position        Joined Loan Date End    Value  \
0           Left   93            RW   Jul 1, 2004           NaN  €103.5M   
1          Right   92            ST  Jul 10, 2018           NaN     €63M   
2          Right   91            GK  Jul 16, 2014           NaN    €120M   
3          Right   91           CAM  Aug 30, 2015           NaN    €129M   
4          Right   91            LW   Aug 3, 2017           NaN    €132M   

    Wage Release Clause  Attacking  Crossing  Finishing  Heading Accuracy  \
0  €560K        €138.4M        429        85         95                70   
1  €220K         €75.9M        437        84         95                90   
2  €125K        €159.4M         95        13         11                15   
3  €370K          €161M        407        94         82                55   
4  €270K        €166.5M        408        85         87                62   

   Short Passing  Volleys  Skill  Dribbling  Curve  FK Accuracy  Long Passing  \
0             91       88    470         96     93           94            91   
1             82       86    414         88     81           76            77   
2             43       13    109         12     13           14            40   
3             94       82    441         88     85           83            93   
4             87       87    448         95     88           89            81   

   Ball Control  Movement  Acceleration  Sprint Speed  Agility  Reactions  \
0            96       451            91            80       91         94   
1            92       431            87            91       87         95   
2            30       307            43            60       67         88   
3            92       398            77            76       78         91   
4            95       453            94            89       96         91   

   Balance  Power  Shot Power  Jumping  Stamina  Strength  Long Shots  \
0       95    389          86       68       72        69          94   
1       71    444          94       95       84        78          93   
2       49    268          59       78       41        78          12   
3       76    408          91       63       89        74          91   
4       83    357          80       62       81        50          84   

   Mentality  Aggression  Interceptions  Positioni

In [3]:
data.tail()

ID          Name        LongName  \
18974  247223        Xia Ao          Ao Xia   
18975  258760      B. Hough       Ben Hough   
18976  252757   R. McKinley  Ronan McKinley   
18977  243790  Wang Zhen'ao    Zhen'ao Wang   
18978  252520     Zhou Xiao       Xiao Zhou   

                                               photoUrl  \
18974  https://cdn.sofifa.com/players/247/223/21_60.png   
18975  https://cdn.sofifa.com/players/258/760/21_60.png   
18976  https://cdn.sofifa.com/players/252/757/21_60.png   
18977  https://cdn.sofifa.com/players/243/790/21_60.png   
18978  https://cdn.sofifa.com/players/252/520/21_60.png   

                                               playerUrl Nationality  Age  \
18974     http://sofifa.com/player/247223/ao-xia/210006/    China PR   21   
18975  http://sofifa.com/player/258760/ben-hough/210006/     England   17   
18976  http://sofifa.com/player/252757/ronan-mckinley...     England   18   
18977  http://sofifa.com/player/243790/zhenao-wang/21...    China PR   20   
18978  http://sofifa.com/player/252520/xiao-zhou/210006/    China PR   21   

       ↓OVA  POT                      Club     Contract Positions Height  \
18974    47   55        \n\n\n\nWuhan Zall  2018 ~ 2022        CB  178cm   
18975    47   67   \n\n\n\nOldham Athletic  2020 ~ 2021        CM  175cm   
18976    47   65        \n\n\n\nDerry City  2019 ~ 2020        CM  179cm   
18977    47   57  \n\n\n\nDalian YiFang FC  2020 ~ 2022        RW  175cm   
18978    47   57  \n\n\n\nDalian YiFang FC  2019 ~ 2023    CB, LB  188cm   

      Weight Preferred Foot  BOV Best Position        Joined Loan Date End  \
18974   66kg          Right   49            CB  Jul 13, 2018           NaN   
18975   65kg          Right   51           CAM   Aug 1, 2020           NaN   
18976   74kg          Right   49           CAM   Mar 8, 2019           NaN   
18977   69kg          Right   48            ST  Sep 22, 2020           NaN   
18978   75kg           Left   50            LB  Jul 29, 2019           NaN   

       Value  Wage Release Clause  Attacking  Crossing  Finishing  \
18974  €100K   €1K           €70K        145        23         26   
18975  €130K  €500          €165K        211        38         42   
18976  €120K  €500          €131K        200        30         34   
18977  €100K   €2K           €88K        215        45         52   
18978  €100K   €1K           €79K        163        40         18   

       Heading Accuracy  Short Passing  Volleys  Skill  Dribbling  Curve  \
18974                43             26       27    142         27     23   
18975                40             56       35    219         46     40   
18976                43             54       39    207         43     39   
18977                34             42       42    194         51     35   
18978                40             45       20    171         40     33   

       FK Accuracy  Long Passing  Ball Control  Movement  Acceleration  \
18974           21            29            42       294            68   
18975           35            50            48       305            63   
18976           31            47            47       290            59   
18977           31            31            46       254            62   
18978           20            40            38       279            60   

       Sprint Speed  Agility  Reactions  Balance  Power  Shot Power  Jumping  \
18974            60       69         46       51    221          36       57   
18975            64       61         51       66    226          48       58   
18976            66       51         47       67    242          45       52   
18977            55       50         33       54    235          56       45   
18978            63       57         47       52    231          37       66   

       Stamina  Strength  Long Shots  Mentality  Aggression  Interceptions  \
18974       54        50          24        192          48             50   
18975       43        47      

## Checking basic info about the data

In [4]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 18979 entries, 0 to 18978
Data columns (total 77 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   ID                18979 non-null  int64 
 1   Name              18979 non-null  object
 2   LongName          18979 non-null  object
 3   photoUrl          18979 non-null  object
 4   playerUrl         18979 non-null  object
 5   Nationality       18979 non-null  object
 6   Age               18979 non-null  int64 
 7   ↓OVA              18979 non-null  int64 
 8   POT               18979 non-null  int64 
 9   Club              18979 non-null  object
 10  Contract          18979 non-null  object
 11  Positions         18979 non-null  object
 12  Height            18979 non-null  object
 13  Weight            18979 non-null  object
 14  Preferred Foot    18979 non-null  object
 15  BOV               18979 non-null  int64 
 16  Best Position     18979 non-null  object
 17  Joined      

There are 18979 rows and 77 columns in our data.

In [5]:
data.duplicated().sum()

0

Our dataset contains no duplicates.

In [6]:
data.columns

Index(['ID', 'Name', 'LongName', 'photoUrl', 'playerUrl', 'Nationality', 'Age',
       '↓OVA', 'POT', 'Club', 'Contract', 'Positions', 'Height', 'Weight',
       'Preferred Foot', 'BOV', 'Best Position', 'Joined', 'Loan Date End',
       'Value', 'Wage', 'Release Clause', 'Attacking', 'Crossing', 'Finishing',
       'Heading Accuracy', 'Short Passing', 'Volleys', 'Skill', 'Dribbling',
       'Curve', 'FK Accuracy', 'Long Passing', 'Ball Control', 'Movement',
       'Acceleration', 'Sprint Speed', 'Agility', 'Reactions', 'Balance',
       'Power', 'Shot Power', 'Jumping', 'Stamina', 'Strength', 'Long Shots',
       'Mentality', 'Aggression', 'Interceptions', 'Positioning', 'Vision',
       'Penalties', 'Composure', 'Defending', 'Marking', 'Standing Tackle',
       'Sliding Tackle', 'Goalkeeping', 'GK Diving', 'GK Handling',
       'GK Kicking', 'GK Positioning', 'GK Reflexes', 'Total Stats',
       'Base Stats', 'W/F', 'SM', 'A/W', 'D/W', 'IR', 'PAC', 'SHO', 'PAS',
       'DRI', 'DEF', 

## Handling Missing Data

In [7]:
#Display the percentage of the missing values for all columns.
pd.options.display.min_rows = 115
percent_missing = data.isnull().sum() * 100 / len(data)
missing_value_data = pd.DataFrame({'percent_missing': percent_missing})
missing_value_data

percent_missing
ID                       0.000000
Name                     0.000000
LongName                 0.000000
photoUrl                 0.000000
playerUrl                0.000000
Nationality              0.000000
Age                      0.000000
↓OVA                     0.000000
POT                      0.000000
Club                     0.000000
Contract                 0.000000
Positions                0.000000
Height                   0.000000
Weight                   0.000000
Preferred Foot           0.000000
BOV                      0.000000
Best Position            0.000000
Joined                   0.000000
Loan Date End           94.662522
Value                    0.000000
Wage                     0.000000
Release Clause           0.000000
Attacking                0.000000
Crossing                 0.000000
Finishing                0.000000
Heading Accuracy         0.000000
Short Passing            0.000000
Volleys                  0.000000
Skill                    0.000000
Dribbling                0.000000
...                           ...
Aggression               0.000000
Interceptions            0.000000
Positioning              0.000000
Vision                   0.000000
Penalties                0.000000
Composure                0.000000
Defending                0.000000
Marking                  0.000000
Standing Tackle          0.000000
Sliding Tackle           0.000000
Goalkeeping              0.000000
GK Diving                0.000000
GK Handling              0.000000
GK Kicking               0.000000
GK Positioning           0.000000
GK Reflexes              0.000000
Total Stats              0.000000
Base Stats               0.000000
W/F                      0.000000
SM                       0.000000
A/W                      0.000000
D/W                      0.000000
IR                       0.000000
PAC                      0.000000
SHO                      0.000000
PAS                      0.000000
DRI                      0.000000
DEF                      0.000000
PHY                      0.000000
Hits                    13.673007

[77 rows x 1 columns]

The Loan Date End and Hits columns both have missing data. Because the Loan Date End column is missing more than 90% of its data, The column will be dropped. The missing values in the Hits column will be replaced with 0.

## Object variable study

In [8]:
data.select_dtypes(object).columns

Index(['Name', 'LongName', 'photoUrl', 'playerUrl', 'Nationality', 'Club',
       'Contract', 'Positions', 'Height', 'Weight', 'Preferred Foot',
       'Best Position', 'Joined', 'Loan Date End', 'Value', 'Wage',
       'Release Clause', 'W/F', 'SM', 'A/W', 'D/W', 'IR', 'Hits'],
      dtype='object')

In [9]:
# Examine the values in 5 rows to determine if they must be integers or if some require a type conversion.

obj_var = data.select_dtypes(object).sample(5)
obj_var.T

11401  \
Name                                                   H. McGahey   
LongName                                         Harrison McGahey   
photoUrl         https://cdn.sofifa.com/players/220/038/21_60.png   
playerUrl       http://sofifa.com/player/220038/harrison-mcgah...   
Nationality                                               England   
Club                                    \n\n\n\nScunthorpe United   
Contract                                              2019 ~ 2021   
Positions                                                      CB   
Height                                                      187cm   
Weight                                                       85kg   
Preferred Foot                                              Right   
Best Position                                                  CB   
Joined                                                Jan 4, 2019   
Loan Date End                                                 NaN   
Value                                                       €750K   
Wage                                                          €4K   
Release Clause                                              €958K   
W/F                                                           3 ★   
SM                                                             2★   
A/W                                                        Medium   
D/W                                                        Medium   
IR                                                            1 ★   
Hits                                                            8   

                                                            11130  \
Name                                                    J. Yeboah   
LongName                                              John Yeboah   
photoUrl         https://cdn.sofifa.com/players/242/755/21_60.png   
playerUrl       http://sofifa.com/player/242755/john-yeboah/21...   
Nationality                                               Germany   
Club                                            \n\n\n\nWillem II   
Contract                                              2020 ~ 2024   
Positions                                                  RW, LW   
Height                                                      170cm   
Weight                                                       66kg   
Preferred Foot                                               Left   
Best Position                                                  LM   
Joined                                               Sep 15, 2020   
Loan Date End                                                 NaN   
Value                                                       €1.5M   
Wage                                                          €2K   
Release Clause                                              €2.1M   
W/F                                                           3 ★   
SM                                                             3★   
A/W                                                        Medium   
D/W                                                        Medium   
IR                                                            1 ★   
Hits                                                           44   

                                                            13043  \
Name                                               Song Jin Hyung   
LongName                                           Jin Hyung Song   
photoUrl         https://cdn.sofifa.com/players/184/334/21_60.png   
playerUrl       http://sofifa.com/player/184334/jin-hyung-song...   
Nationality                                        Korea Republic   
Club                                             \n\n\n\nFC Seoul   
Contract                                              2017 ~ 2023   
Positions                                            CM, CAM, CDM   
Height                                                      175cm   
Weight                                                  

## STEPS FOR CLEANING

1. Rename columns

2. Remove the 'cm' and 'kg' from the weight and height columns and check for other signs and convert to the appropriate one.

3. Delete the '€', 'M', and 'K' symbols from the Value, Wage, and Release Clause columns and convert them to appropriate values.

4. Remove the '★' sign from the W/F, SM, and IR columns.

5. Examine the Hits column for any more anomalies.

6. Clean the contract column

7. Convert the joined column to a datetime column.

8. Clean club column

9. Drop unnecessary columns

## 1. Rename columns

In [10]:
# Replace columns having space inbetween with '_'
data.columns = [i.replace(' ', '_') for i in data.columns]

## 2. Dealing with the Weight and Height Columns

### Height Column

In [11]:
#Check if all the heights are in 'cm'

data['Height'][data['Height'].notnull()].str.endswith('cm').all()


False

This means some values are not in 'cm' so let's find out the others.

In [12]:
data['Height'].unique()

array(['170cm', '187cm', '188cm', '181cm', '175cm', '184cm', '191cm',
       '178cm', '193cm', '185cm', '199cm', '173cm', '168cm', '176cm',
       '177cm', '183cm', '180cm', '189cm', '179cm', '195cm', '172cm',
       '182cm', '186cm', '192cm', '165cm', '194cm', '167cm', '196cm',
       '163cm', '190cm', '174cm', '169cm', '171cm', '197cm', '200cm',
       '166cm', '6\'2"', '164cm', '198cm', '6\'3"', '6\'5"', '5\'11"',
       '6\'4"', '6\'1"', '6\'0"', '5\'10"', '5\'9"', '5\'6"', '5\'7"',
       '5\'4"', '201cm', '158cm', '162cm', '161cm', '160cm', '203cm',
       '157cm', '156cm', '202cm', '159cm', '206cm', '155cm'], dtype=object)

There are some heights measured in inches (") as well. Everything will be converted to 'cm'

In [13]:
# define a function to convert height from feet and inches to cm.
def convert_height(y):
    
    # Split the height string into foot and inches parts
    if y[-1] == '"':
        y = y.replace('\"', '')
        inch = int(y[2:])
        foot = int(y[0])
        
    # Convert foot and inches to centimeters
        total_inches = foot * 12 + inch
        cm = total_inches * 2.54
        return round(cm)
    elif y[-1] == "m":
        return int(y[:-2])
    

#Apply function to Height column
data['Height'] = data['Height'].apply(convert_height)

In [14]:
data['Height'].unique()

array([170, 187, 188, 181, 175, 184, 191, 178, 193, 185, 199, 173, 168,
       176, 177, 183, 180, 189, 179, 195, 172, 182, 186, 192, 165, 194,
       167, 196, 163, 190, 174, 169, 171, 197, 200, 166, 164, 198, 201,
       158, 162, 161, 160, 203, 157, 156, 202, 159, 206, 155], dtype=int64)

### Weight Column

In [15]:
#Check if all the weight are in 'kg'

data['Weight'][data['Weight'].notnull()].str.endswith('kg').all()

False

In [16]:
data['Weight'].unique()

array(['72kg', '83kg', '87kg', '70kg', '68kg', '80kg', '71kg', '91kg',
       '73kg', '85kg', '92kg', '69kg', '84kg', '96kg', '81kg', '82kg',
       '75kg', '86kg', '89kg', '74kg', '76kg', '64kg', '78kg', '90kg',
       '66kg', '60kg', '94kg', '79kg', '67kg', '65kg', '59kg', '61kg',
       '93kg', '88kg', '97kg', '77kg', '62kg', '63kg', '95kg', '100kg',
       '58kg', '183lbs', '179lbs', '172lbs', '196lbs', '176lbs', '185lbs',
       '170lbs', '203lbs', '168lbs', '161lbs', '146lbs', '130lbs',
       '190lbs', '174lbs', '148lbs', '165lbs', '159lbs', '192lbs',
       '181lbs', '139lbs', '154lbs', '157lbs', '163lbs', '98kg', '103kg',
       '99kg', '102kg', '56kg', '101kg', '57kg', '55kg', '104kg', '107kg',
       '110kg', '53kg', '50kg', '54kg', '52kg'], dtype=object)

Some weights were measured in lbs so they'll be converted to kg

In [17]:
# define a function to convert weight from lbs to kg.
def convert_weight(y):   
    if y[-1] == 's':
        y = y.replace('lbs', '')
        lbs = int(y[:3])
        
    # Convert lbs to kg
        kg = lbs * 0.45359237
        return round(kg)
    elif y[-1] == "g":
        return int(y[:-2])
    

#Apply function to Height column
data['Weight'] = data['Weight'].apply(convert_weight)

In [18]:
data['Weight'].unique()

array([ 72,  83,  87,  70,  68,  80,  71,  91,  73,  85,  92,  69,  84,
        96,  81,  82,  75,  86,  89,  74,  76,  64,  78,  90,  66,  60,
        94,  79,  67,  65,  59,  61,  93,  88,  97,  77,  62,  63,  95,
       100,  58,  98, 103,  99, 102,  56, 101,  57,  55, 104, 107, 110,
        53,  50,  54,  52], dtype=int64)

## 3. Removing the '€', 'M', and 'K' symbols and converting them to appropriate values.

In [19]:
#Check if the '€' is is the only currency in the columns

value = data['Value']
wage = data['Wage']
release_clause = data['Release_Clause']

In [20]:
value.str[0].unique()

array(['€'], dtype=object)

In [21]:
wage.str[0].unique()

array(['€'], dtype=object)

In [22]:
release_clause.str[0].unique()

array(['€'], dtype=object)

They all have the Euro sign as the only unique sign

In [23]:
# define a function to convert value to float
def convert_value(val):
    if val[-1] == 'M':
        return float(val[1:-1]) * 1000000
    elif val[-1] == 'K':
        return float(val[1:-1]) * 1000
    else:
        return float(val[1:])

# apply function to value column
data['Value'] = data['Value'].apply(convert_value)
data['Wage'] = data['Wage'].apply(convert_value)
data['Release_Clause'] = data['Release_Clause'].apply(convert_value)

data.head()

ID               Name                      LongName  \
0  158023           L. Messi                  Lionel Messi   
1   20801  Cristiano Ronaldo  C. Ronaldo dos Santos Aveiro   
2  200389           J. Oblak                     Jan Oblak   
3  192985       K. De Bruyne               Kevin De Bruyne   
4  190871          Neymar Jr    Neymar da Silva Santos Jr.   

                                           photoUrl  \
0  https://cdn.sofifa.com/players/158/023/21_60.png   
1  https://cdn.sofifa.com/players/020/801/21_60.png   
2  https://cdn.sofifa.com/players/200/389/21_60.png   
3  https://cdn.sofifa.com/players/192/985/21_60.png   
4  https://cdn.sofifa.com/players/190/871/21_60.png   

                                           playerUrl Nationality  Age  ↓OVA  \
0  http://sofifa.com/player/158023/lionel-messi/2...   Argentina   33    93   
1  http://sofifa.com/player/20801/c-ronaldo-dos-s...    Portugal   35    92   
2  http://sofifa.com/player/200389/jan-oblak/210006/    Slovenia   27    91   
3  http://sofifa.com/player/192985/kevin-de-bruyn...     Belgium   29    91   
4  http://sofifa.com/player/190871/neymar-da-silv...      Brazil   28    91   

   POT                         Club     Contract   Positions  Height  Weight  \
0   93         \n\n\n\nFC Barcelona  2004 ~ 2021  RW, ST, CF     170      72   
1   92             \n\n\n\nJuventus  2018 ~ 2022      ST, LW     187      83   
2   93      \n\n\n\nAtlético Madrid  2014 ~ 2023          GK     188      87   
3   91      \n\n\n\nManchester City  2015 ~ 2023     CAM, CM     181      70   
4   91  \n\n\n\nParis Saint-Germain  2017 ~ 2022     LW, CAM     175      68   

  Preferred_Foot  BOV Best_Position        Joined Loan_Date_End        Value  \
0           Left   93            RW   Jul 1, 2004           NaN  103500000.0   
1          Right   92            ST  Jul 10, 2018           NaN   63000000.0   
2          Right   91            GK  Jul 16, 2014           NaN  120000000.0   
3          Right   91           CAM  Aug 30, 2015           NaN  129000000.0   
4          Right   91            LW   Aug 3, 2017           NaN  132000000.0   

       Wage  Release_Clause  Attacking  Crossing  Finishing  Heading_Accuracy  \
0  560000.0     138400000.0        429        85         95                70   
1  220000.0      75900000.0        437        84         95                90   
2  125000.0     159400000.0         95        13         11                15   
3  370000.0     161000000.0        407        94         82                55   
4  270000.0     166500000.0        408        85         87                62   

   Short_Passing  Volleys  Skill  Dribbling  Curve  FK_Accuracy  Long_Passing  \
0             91       88    470         96     93           94            91   
1             82       86    414         88     81           76            77   
2             43       13    109         12     13           14            40   
3             94       82    441         88     85           83            93   
4             87       87    448         95     88           89            81   

   Ball_Control  Movement  Acceleration  Sprint_Speed  Agility  Reactions  \
0            96       451            91            80       91         94   
1            92       431            87            91       87         95   
2            30       307            43            60       67         88   
3            92       398            77            76       78         91   
4            95       453            94            89       96         91   

   Balance  Power  Shot_Power  Jumping  Stamina  Strength  Long_Shots  \
0       95    389          86       68       72        69          94   
1       71    444          94       95       84        78          93   
2       49    268          59       78       41        78          12   
3       76    408          91       63       89        74          91   
4       83    357          80       62       81        50       

## 4. Remove the '★' sign from the W/F, SM, and IR columns.

In [24]:
data['W/F'].unique()

array(['4 ★', '3 ★', '5 ★', '2 ★', '1 ★'], dtype=object)

In [25]:
data['SM'].unique()

array(['4★', '5★', '1★', '2★', '3★'], dtype=object)

In [26]:
data['IR'].unique()

array(['5 ★', '3 ★', '4 ★', '2 ★', '1 ★'], dtype=object)

In [27]:
# define a function to remove ★
def remove_star(val):
    if val[-1] == '★':
        return (val[:-1]) 
    else:
        return (val)

# apply function to value column
data['SM'] = data['SM'].apply(remove_star)
data['IR'] = data['IR'].apply(remove_star)
data['W/F'] = data['W/F'].apply(remove_star)


data.head()

ID               Name                      LongName  \
0  158023           L. Messi                  Lionel Messi   
1   20801  Cristiano Ronaldo  C. Ronaldo dos Santos Aveiro   
2  200389           J. Oblak                     Jan Oblak   
3  192985       K. De Bruyne               Kevin De Bruyne   
4  190871          Neymar Jr    Neymar da Silva Santos Jr.   

                                           photoUrl  \
0  https://cdn.sofifa.com/players/158/023/21_60.png   
1  https://cdn.sofifa.com/players/020/801/21_60.png   
2  https://cdn.sofifa.com/players/200/389/21_60.png   
3  https://cdn.sofifa.com/players/192/985/21_60.png   
4  https://cdn.sofifa.com/players/190/871/21_60.png   

                                           playerUrl Nationality  Age  ↓OVA  \
0  http://sofifa.com/player/158023/lionel-messi/2...   Argentina   33    93   
1  http://sofifa.com/player/20801/c-ronaldo-dos-s...    Portugal   35    92   
2  http://sofifa.com/player/200389/jan-oblak/210006/    Slovenia   27    91   
3  http://sofifa.com/player/192985/kevin-de-bruyn...     Belgium   29    91   
4  http://sofifa.com/player/190871/neymar-da-silv...      Brazil   28    91   

   POT                         Club     Contract   Positions  Height  Weight  \
0   93         \n\n\n\nFC Barcelona  2004 ~ 2021  RW, ST, CF     170      72   
1   92             \n\n\n\nJuventus  2018 ~ 2022      ST, LW     187      83   
2   93      \n\n\n\nAtlético Madrid  2014 ~ 2023          GK     188      87   
3   91      \n\n\n\nManchester City  2015 ~ 2023     CAM, CM     181      70   
4   91  \n\n\n\nParis Saint-Germain  2017 ~ 2022     LW, CAM     175      68   

  Preferred_Foot  BOV Best_Position        Joined Loan_Date_End        Value  \
0           Left   93            RW   Jul 1, 2004           NaN  103500000.0   
1          Right   92            ST  Jul 10, 2018           NaN   63000000.0   
2          Right   91            GK  Jul 16, 2014           NaN  120000000.0   
3          Right   91           CAM  Aug 30, 2015           NaN  129000000.0   
4          Right   91            LW   Aug 3, 2017           NaN  132000000.0   

       Wage  Release_Clause  Attacking  Crossing  Finishing  Heading_Accuracy  \
0  560000.0     138400000.0        429        85         95                70   
1  220000.0      75900000.0        437        84         95                90   
2  125000.0     159400000.0         95        13         11                15   
3  370000.0     161000000.0        407        94         82                55   
4  270000.0     166500000.0        408        85         87                62   

   Short_Passing  Volleys  Skill  Dribbling  Curve  FK_Accuracy  Long_Passing  \
0             91       88    470         96     93           94            91   
1             82       86    414         88     81           76            77   
2             43       13    109         12     13           14            40   
3             94       82    441         88     85           83            93   
4             87       87    448         95     88           89            81   

   Ball_Control  Movement  Acceleration  Sprint_Speed  Agility  Reactions  \
0            96       451            91            80       91         94   
1            92       431            87            91       87         95   
2            30       307            43            60       67         88   
3            92       398            77            76       78         91   
4            95       453            94            89       96         91   

   Balance  Power  Shot_Power  Jumping  Stamina  Strength  Long_Shots  \
0       95    389          86       68       72        69          94   
1       71    444          94       95       84        78          93   
2       49    268          59       78       41        78          12   
3       76    408          91       63       89        74          91   
4       83    357          80       62       81        50       

In [28]:
#Convert them to integert values
data['W/F'] = data['W/F'].astype(int)
data['IR'] = data['IR'].astype(int)
data['SM'] = data['SM'].astype(int)

## 5. Examine the Hits column for any more anomalies.

In [29]:
data['Hits'].unique()

array(['771', '562', '150', '207', '595', '248', '246', '120', '1.6K',
       '130', '321', '189', '175', '96', '118', '216', '212', '154',
       '205', '202', '339', '408', '103', '332', '86', '173', '161',
       '396', '1.1K', '433', '242', '206', '177', '1.5K', '198', '459',
       '117', '119', '209', '84', '187', '165', '203', '65', '336', '126',
       '313', '124', '145', '538', '182', '101', '45', '377', '99', '194',
       '403', '414', '593', '374', '245', '3.2K', '266', '299', '309',
       '215', '265', '211', '112', '337', '70', '159', '688', '116', '63',
       '144', '123', '71', '224', '113', '168', '61', '89', '137', '278',
       '75', '148', '176', '197', '264', '214', '247', '402', '440',
       '1.7K', '2.3K', '171', '320', '657', '87', '259', '200', '255',
       '253', '196', '60', '97', '85', '169', '256', '132', '239', '166',
       '121', '109', '32', '46', '122', '48', '527', '199', '282', '51',
       '1.9K', '642', '155', '323', '288', '497', '509', '79',

Remember that the Hits column has missing values. We'll replace the null values with 0.

In [30]:
# replace null values with 0
data['Hits'].fillna('0', inplace=True)

In [31]:
#define a function to replace the 'K' with 1000
def hits_col(val):
    val = val.lstrip()
    if val.endswith('K'):
        val = val[:-1]
        val = float(val) * 1000
        val = int(val)
    else:
        val = int(val)
    return val

data['Hits'] = data['Hits'].apply(hits_col)

In [32]:
data['Hits'].dtype

dtype('int64')

## 6. Clean the contract column

In [33]:
data['Contract'].unique()

array(['2004 ~ 2021', '2018 ~ 2022', '2014 ~ 2023', '2015 ~ 2023',
       '2017 ~ 2022', '2017 ~ 2023', '2018 ~ 2024', '2014 ~ 2022',
       '2018 ~ 2023', '2016 ~ 2023', '2013 ~ 2023', '2011 ~ 2023',
       '2009 ~ 2022', '2005 ~ 2021', '2011 ~ 2021', '2015 ~ 2022',
       '2017 ~ 2024', '2010 ~ 2024', '2012 ~ 2021', '2019 ~ 2024',
       '2015 ~ 2024', '2017 ~ 2025', '2020 ~ 2025', '2019 ~ 2023',
       '2008 ~ 2023', '2015 ~ 2021', '2020 ~ 2022', '2012 ~ 2022',
       '2016 ~ 2025', '2013 ~ 2022', '2011 ~ 2022', '2012 ~ 2024',
       '2016 ~ 2021', '2012 ~ 2023', '2008 ~ 2022', '2019 ~ 2022',
       '2017 ~ 2021', '2013 ~ 2024', '2020 ~ 2024', '2010 ~ 2022',
       '2020 ~ 2021', '2011 ~ 2024', '2020 ~ 2023', '2014 ~ 2024',
       '2013 ~ 2026', '2016 ~ 2022', '2010 ~ 2021', '2013 ~ 2021',
       '2019 ~ 2025', '2018 ~ 2025', '2016 ~ 2024', '2018 ~ 2021',
       '2009 ~ 2024', '2007 ~ 2022', 'Jun 30, 2021 On Loan',
       '2009 ~ 2021', '2019 ~ 2021', '2019 ~ 2026', 'Free', '2012 ~ 

Looks like Some players are on contracts while others are on Loan. The column will be splitted and to have the start and end of the contract.

In [34]:
contract = data['Contract'].copy()

In [35]:
#define a function to for the start and end dates of contract column.
def contract_date(val):
  
    if '~' in val:
        dates = val.split(' ~ ')
        start_year = dates[0]
        start_year = dates[1]
        return [start_year, start_year]

    else:
        return 'NAN'   

In [36]:
contract = contract.apply(contract_date)

In [37]:
data['start_date'] = [i[0] if len(i[0]) == 4 else np.nan for i in contract]

In [38]:
data['end_date'] = [i[1] if len(i[0]) == 4 else np.nan for i in contract]

In [39]:
data['end_date'].unique()

array(['2021', '2022', '2023', '2024', '2025', '2026', nan, '2028',
       '2020', '2027'], dtype=object)

In [40]:
# Add a placeholder month and day to each year, and convert to datetime
data['start_date'] = pd.to_datetime(data['start_date'] + '-01-01', format='%Y-%m-%d', errors='coerce')
data['end_date'] = pd.to_datetime(data['end_date']+ '-01-01', format='%Y-%m-%d', errors='coerce')

#df['Year'] = pd.to_datetime(df['Year'], errors='coerce')


In [41]:
data['end_date'].dtype

dtype('<M8[ns]')

## 7. Convert the joined column to a datetime column.

In [42]:
# convert date column to datetime format
data['Joined'] = pd.to_datetime(data['Joined'], format='%b %d, %Y')

# extract the day of the month from the datetime column
data['Day'] = data['Joined'].dt.day

# extract the month from the datetime column as a name or number
data['Month_name'] = data['Joined'].dt.month_name()
data['Month_number'] = data['Joined'].dt.month

# extract the year from the datetime column
data['Year'] = data['Joined'].dt.year

data.head()

ID               Name                      LongName  \
0  158023           L. Messi                  Lionel Messi   
1   20801  Cristiano Ronaldo  C. Ronaldo dos Santos Aveiro   
2  200389           J. Oblak                     Jan Oblak   
3  192985       K. De Bruyne               Kevin De Bruyne   
4  190871          Neymar Jr    Neymar da Silva Santos Jr.   

                                           photoUrl  \
0  https://cdn.sofifa.com/players/158/023/21_60.png   
1  https://cdn.sofifa.com/players/020/801/21_60.png   
2  https://cdn.sofifa.com/players/200/389/21_60.png   
3  https://cdn.sofifa.com/players/192/985/21_60.png   
4  https://cdn.sofifa.com/players/190/871/21_60.png   

                                           playerUrl Nationality  Age  ↓OVA  \
0  http://sofifa.com/player/158023/lionel-messi/2...   Argentina   33    93   
1  http://sofifa.com/player/20801/c-ronaldo-dos-s...    Portugal   35    92   
2  http://sofifa.com/player/200389/jan-oblak/210006/    Slovenia   27    91   
3  http://sofifa.com/player/192985/kevin-de-bruyn...     Belgium   29    91   
4  http://sofifa.com/player/190871/neymar-da-silv...      Brazil   28    91   

   POT                         Club     Contract   Positions  Height  Weight  \
0   93         \n\n\n\nFC Barcelona  2004 ~ 2021  RW, ST, CF     170      72   
1   92             \n\n\n\nJuventus  2018 ~ 2022      ST, LW     187      83   
2   93      \n\n\n\nAtlético Madrid  2014 ~ 2023          GK     188      87   
3   91      \n\n\n\nManchester City  2015 ~ 2023     CAM, CM     181      70   
4   91  \n\n\n\nParis Saint-Germain  2017 ~ 2022     LW, CAM     175      68   

  Preferred_Foot  BOV Best_Position     Joined Loan_Date_End        Value  \
0           Left   93            RW 2004-07-01           NaN  103500000.0   
1          Right   92            ST 2018-07-10           NaN   63000000.0   
2          Right   91            GK 2014-07-16           NaN  120000000.0   
3          Right   91           CAM 2015-08-30           NaN  129000000.0   
4          Right   91            LW 2017-08-03           NaN  132000000.0   

       Wage  Release_Clause  Attacking  Crossing  Finishing  Heading_Accuracy  \
0  560000.0     138400000.0        429        85         95                70   
1  220000.0      75900000.0        437        84         95                90   
2  125000.0     159400000.0         95        13         11                15   
3  370000.0     161000000.0        407        94         82                55   
4  270000.0     166500000.0        408        85         87                62   

   Short_Passing  Volleys  Skill  Dribbling  Curve  FK_Accuracy  Long_Passing  \
0             91       88    470         96     93           94            91   
1             82       86    414         88     81           76            77   
2             43       13    109         12     13           14            40   
3             94       82    441         88     85           83            93   
4             87       87    448         95     88           89            81   

   Ball_Control  Movement  Acceleration  Sprint_Speed  Agility  Reactions  \
0            96       451            91            80       91         94   
1            92       431            87            91       87         95   
2            30       307            43            60       67         88   
3            92       398            77            76       78         91   
4            95       453            94            89       96         91   

   Balance  Power  Shot_Power  Jumping  Stamina  Strength  Long_Shots  \
0       95    389          86       68       72        69          94   
1       71    444          94       95       84        78          93   
2       49    268          59       78       41        78          12   
3       76    408          91       63       89        74          91   
4       83    357          80       62       81        50          84   

   Menta

## 8. Clean Club column

In [44]:
data.Club.sample(10)

17545       \n\n\n\nFC Chambly Oise
7385         \n\n\n\nStade de Reims
17312         \n\n\n\nReal Sociedad
3326            \n\n\n\nKV Kortrijk
9367       \n\n\n\nPortland Timbers
2476         \n\n\n\nGalatasaray SK
17635          \n\n\n\nAalesunds FK
18383    \n\n\n\nArgentinos Juniors
16238     \n\n\n\nBayern München II
4661          \n\n\n\nGórnik Zabrze
Name: Club, dtype: object

In [45]:
# Remove \n for begining of club names
clubs = data['Club'].str.strip()
clubs.head()

0           FC Barcelona
1               Juventus
2        Atlético Madrid
3        Manchester City
4    Paris Saint-Germain
Name: Club, dtype: object

I discovered that some club names had a number at the beginning. I will verify before I take the numbers off.

In [46]:
clubs[clubs.str.startswith('1. ')].count()

234

In [47]:
#Removing '1.'' from all of them
clubs = clubs.str.lstrip('1. ')

In [48]:
clubs[clubs.str.startswith('1. ')].count()

0

In [49]:
clubs.sample(10)

1683                  Lazio
8119           Nashville SC
3158     Argentinos Juniors
6618               Al Hazem
2245         Spartak Moscow
2185       Atlético Mineiro
16703           Club Necaxa
9956          Club Blooming
14674     Oriente Petrolero
16488     Bayern München II
Name: Club, dtype: object

Also, there were some rows in the column that had no club. They will be replaced with NaN values

In [50]:
clubs[clubs == 'No Club'].count()

237

In [51]:
clubs[clubs == 'No Club'] = np.nan

In [52]:
#Set the cleaned clubs values to the club column in the dataset 
data['Club'] = clubs

## 9. Drop unnecessary columns


The "Positions" column was removed since each player's "Best Position" is listed in a separate column. For players who had several positions, the "Positions" column also included the other positions in addition to the information from the "Best Position" column.

Both the "playerUrl" and the "photoUrl" include data that are not required for analysis.

The contract column has been split to start and end dates so the column will no longer be necessary

The "Loan Date End" column will also be removed because it primarily includes null values.

In [53]:
data.drop(['photoUrl', 'playerUrl', 'Loan_Date_End', 'Contract', 'Positions'], axis = 1, inplace = True)

In [54]:
# Examine the values in 5 rows to determine if they must be integers or if some require a type conversion.

obj_var = data.select_dtypes(object).sample(5)
obj_var.T

3142                  15950        13914  \
Name                        J. Durmaz     Y. Schoonderwaldt    C. Coskun   
LongName                 Jimmy Durmaz  Youri Schoonderwaldt   Can Coskun   
Nationality                    Sweden           Netherlands       Turkey   
Club            Fatih Karagümrük S.K.          ADO Den Haag  FSV Zwickau   
Preferred_Foot                   Left                 Right         Left   
Best_Position                     CAM                    GK           LM   
A/W                            Medium                Medium         High   
D/W                               Low                Medium         High   
Month_name                       July                August      January   

                                       5436                 15026  
Name                          Villiam Moraes          Y. Schlößer  
LongName        Villiam Carlos Moraes Santos     Yannick Schlößer  
Nationality                           Brazil              Germany  
Club                             Oceânico FC  Bayer 04 Leverkusen  
Preferred_Foot                          Left                Right  
Best_Position                             CB                   ST  
A/W                                   Medium               Medium  
D/W                                   Medium               Medium  
Month_name                           January                 July